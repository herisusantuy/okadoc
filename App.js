import React, {useState} from 'react';
import {Text, View, Image, TextInput, ScrollView} from 'react-native';
import CustomButton from './js/component/common/CustomButton';
import {
  multiply,
  add,
  getPrime,
  getFibonacci,
  isNumeric,
} from './js/helper/formula';
import {colors} from './js/styles/default';
import styles from './js/styles/app';
const buttons = [
  'Multiply X & Y',
  'Sum X & Y',
  'First N Prime',
  'First N Fibonacci',
];
const App = () => {
  const [value, setValue] = useState(null);
  const [result, setResult] = useState(null);
  const [errMsg, setErrMsg] = useState(null);
  const [resultTitle, setResultTitle] = useState(null);
  const [button, setButton] = useState(null);

  const validate = button => {
    if (!value) {
      setErrMsg(`You haven't input any number!`);
      return false;
    } else {
      let inputArr = value.trim().split(',');

      if (button == 'First N Prime' || button == 'First N Fibonacci') {
        if (!isNumeric(inputArr[0])) {
          setErrMsg(`Please input Number!`);
          return false;
        }
        if (inputArr.length > 1) {
          setErrMsg(`Please input positive number!`);
          return false;
        }
      } else {
        if (inputArr.length == 1 || inputArr.length > 2) {
          setErrMsg(`Please input 2 numbers separated by coma(,)!`);
          return false;
        }
        if (!isNumeric(inputArr[0]) || !isNumeric(inputArr[1])) {
          setErrMsg(`Please input Number than String!`);
          return false;
        }
      }
    }
    return true;
  };

  const calculate = button => {
    setResult(null);
    setErrMsg(null);
    if (validate(button)) {
      setButton(button);
      let input = value.trim().split(',');
      if (button.toLowerCase().includes('multiply')) {
        setResult(multiply(input[0], input[1]));
        setResultTitle(
          `Here's the result  multiply ${input[0]} and ${input[1]} is:`,
        );
      }
      if (button.toLowerCase().includes('sum')) {
        setResult(add(Number(input[0]), Number(input[1])));
        setResultTitle(
          `Here's the result  sum ${input[0]} and ${input[1]} is:`,
        );
      }
      if (button.toLowerCase().includes('prime')) {
        setResult(getPrime(Number(input[0])));
        setResultTitle(
          `Here's the result  first N(${input[0]}) Prime numbers are:`,
        );
      }
      if (button.toLowerCase().includes('fibonacci')) {
        setResult(getFibonacci(Number(input[0])));
        setResultTitle(
          `Here's the result  first N(${input[0]}) Fibonacci sequence are:`,
        );
      }
    }
  };

  const renderButtons = buttons.map((button, i) => (
    <CustomButton
      key={i}
      title={button}
      onPressButton={() => calculate(button)}
    />
  ));
  return (
    <View style={styles.container}>
      <View style={[styles.titleContainer]}>
        <Image
          style={styles.logo}
          source={require('./js/assets/okadoc_logo.png')}
        />
        <Text style={styles.title}>Technical Test</Text>
      </View>
      <View style={[styles.inputContainer, {justifyContent: 'flex-end'}]}>
        {result && (
          <View style={[styles.resultContainer]}>
            <Text style={styles.resultTitle}>{resultTitle}</Text>
            <ScrollView>
              <Text
                style={[
                  styles.resultText,
                  {
                    color:
                      button == 'Multiply X & Y'
                        ? colors.green
                        : button == 'Sum X & Y'
                        ? colors.blue
                        : button == 'First N Prime'
                        ? colors.orange
                        : colors.red,
                    fontSize:
                      button == 'Multiply X & Y' || button == 'Sum X & Y'
                        ? 100
                        : 20,
                  },
                ]}>
                {result}
              </Text>
            </ScrollView>
          </View>
        )}
        {!errMsg && !result && (
          <View style={styles.doctorContainer}>
            <Image
              style={styles.doctor}
              source={require('./js/assets/doctor.png')}
            />
          </View>
        )}
        {errMsg && (
          <View style={[styles.resultContainer]}>
            <Text style={styles.errorText}>{errMsg}</Text>
          </View>
        )}
        <TextInput
          placeholder="Please input your number..."
          style={styles.input}
          value={value}
          onChangeText={value => {
            setValue(value);
            setErrMsg(null);
            setResult(null);
          }}
        />
      </View>
      <View style={styles.buttonContainer}>{renderButtons}</View>
    </View>
  );
};

export default App;
