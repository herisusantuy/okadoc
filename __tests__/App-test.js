import 'react-native';
import React from 'react';
import App from '../App';
import CustomeButton from '../js/component/common/CustomButton';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

test('App snapShot', () => {
  const snap = renderer.create(<App />).toJSON();
  expect(snap).toMatchSnapshot();
});

test('Custom Button snapShot', () => {
  const snap = renderer.create(<CustomeButton />).toJSON();
  expect(snap).toMatchSnapshot();
});
