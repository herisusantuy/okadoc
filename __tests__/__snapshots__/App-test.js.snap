// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapShot 1`] = `
<View
  style={
    Object {
      "marginTop": 28.115942028985508,
      "paddingHorizontal": 14.057971014492754,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "height": 281.15942028985506,
          "justifyContent": "center",
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../js/assets/okadoc_logo.png",
        }
      }
      style={
        Object {
          "height": 126.52173913043478,
          "resizeMode": "center",
          "width": "90%",
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#52C7D5",
          "fontSize": 33.73913043478261,
          "fontWeight": "bold",
          "marginBottom": 21.086956521739133,
          "textAlign": "center",
        }
      }
    >
      Technical Test
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "height": 421.7391304347826,
          "justifyContent": "space-between",
          "paddingHorizontal": 21.086956521739133,
        },
        Object {
          "justifyContent": "flex-end",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "justifyContent": "center",
          "paddingBottom": 21.086956521739133,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../js/assets/doctor.png",
          }
        }
        style={
          Object {
            "height": "95%",
            "resizeMode": "center",
            "width": "100%",
          }
        }
      />
    </View>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Please input your number..."
      rejectResponderTermination={true}
      style={
        Object {
          "borderColor": "skyblue",
          "borderRadius": 7.028985507246377,
          "borderWidth": 1,
          "padding": 14.057971014492754,
        }
      }
      underlineColorAndroid="transparent"
      value={null}
    />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "flexWrap": "wrap",
        "height": 421.7391304347826,
        "justifyContent": "center",
        "marginVertical": 28.115942028985508,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#186C62",
          "borderRadius": 7.028985507246377,
          "height": 70.28985507246377,
          "justifyContent": "center",
          "marginRight": 7.028985507246377,
          "marginVertical": 7.028985507246377,
          "opacity": 1,
          "width": "45%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontWeight": "bold",
          }
        }
      >
        Multiply X & Y
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#1361A2",
          "borderRadius": 7.028985507246377,
          "height": 70.28985507246377,
          "justifyContent": "center",
          "marginRight": 7.028985507246377,
          "marginVertical": 7.028985507246377,
          "opacity": 1,
          "width": "45%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontWeight": "bold",
          }
        }
      >
        Sum X & Y
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FCC004",
          "borderRadius": 7.028985507246377,
          "height": 70.28985507246377,
          "justifyContent": "center",
          "marginRight": 7.028985507246377,
          "marginVertical": 7.028985507246377,
          "opacity": 1,
          "width": "45%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontWeight": "bold",
          }
        }
      >
        First N Prime
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#F80701",
          "borderRadius": 7.028985507246377,
          "height": 70.28985507246377,
          "justifyContent": "center",
          "marginRight": 7.028985507246377,
          "marginVertical": 7.028985507246377,
          "opacity": 1,
          "width": "45%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontWeight": "bold",
          }
        }
      >
        First N Fibonacci
      </Text>
    </View>
  </View>
</View>
`;

exports[`Custom Button snapShot 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F80701",
      "borderRadius": 7.028985507246377,
      "height": 70.28985507246377,
      "justifyContent": "center",
      "marginRight": 7.028985507246377,
      "marginVertical": 7.028985507246377,
      "opacity": 1,
      "width": "45%",
    }
  }
>
  <Text
    style={
      Object {
        "color": "#fff",
        "fontWeight": "bold",
      }
    }
  />
</View>
`;
